### PR TITLE
fix: Be more sensible when adding changes to list of pending changes

### PR DESCRIPTION
**Pull request by TatuLund**
Fixes: https://github.com/vaadin/flow/issues/13851

### CODE REVIEW

**Review by MarcinVaadin — APPROVED** [review]
Changes reviewed and tested, memory usage stays on same level.

**Review by mshabarov — APPROVED** [review]
Looks good to me.
This patch re-indexes changes once any removal change is added, this gives a memory consumption improvement: removed items are eligible for garbage collection early, not only after sending a response to client.
Additionally, the logic and codes in `addChange` are simpler, than were before in `collectChanges` in my opinion.

I made a quick performance test using the example project from the issue's author (modified it a bit to include removals):

```java
        add(new Button("Click me", e -> {
            IntStream.range(0, 10000).forEach(i -> {
                testLayout.add(new Button("B"));

                if (i > 5000 && random.nextBoolean()) {
                    int componentCount = testLayout.getComponentCount();
                    testLayout.remove(testLayout.getComponentAt(
                            random.nextInt(componentCount - 1)));
                }

            });
        }));
```

New algorithm has more traverses/loops and these traverses happen on each removal, which makes the overall algorithm complexity higher. However, I believe that it shouldn't affect the performance, until one has a millions of elements in the change list - not sure it's a case for real applications. 
The test gives me approximately the same response processing time as before the patch.

File: flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/NodeList.java [hunk 1]
```diff
@@ -277,59 +275,88 @@ protected List<AbstractListChange<T>> getChangeTracker() {
...
-                        : "AbstractListChange has only three subtypes: add, remove and clear";
-            }
-            index++;
-        }
-
-        List<AbstractListChange<T>> changes;
-
-        changes = allChanges.stream().filter(this::acceptChange)
-                .collect(Collectors.toList());
+        List<AbstractListChange<T>> changes = getChangeTracker().stream()
```
Comment (knoobie) [inline flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/NodeList.java:358]: If you worried about performance of the added loops, you can refactor this stream usage to a loop to improve performance :) 

See e.g. https://medium.com/levi-niners-crafts/java-performance-improvement-java-8-streams-vs-loops-and-lists-vs-arrays-e824136832d6 and hundred other threads about loops > sequential streams performance vise

Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/NodeList.java]: Thanks for the tip @knoobie . I made a quick test stream + filter vs loop and indeed it gives me around 3-4 times faster results for loops while testing with N ~1000 of items.

**Review by mshabarov — APPROVED** [review]

**Comment by ZheSun88** [comment]
2.8 CP -> https://github.com/vaadin/flow/pull/14586

**Pull request merged**